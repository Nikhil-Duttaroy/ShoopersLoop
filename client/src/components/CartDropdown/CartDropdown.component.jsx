import React from "react";
import { connect,useSelector} from "react-redux";
import CartItem from "./../CartItem/CartItem.component";

import { createStructuredSelector } from "reselect";
import { withRouter } from "react-router-dom";

import { selectCartItems } from "./../../redux/cart/cart.selector";
import { toggleCartHidden } from "../../redux/cart/cart.actions";



import {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer,
} from "./CartDropdown.styles";

export const CartDropdown = ({ cartItems, history, dispatch }) => {
  const user=useSelector(state=>state.user.currentUser)
  // console.log(user);
  return (
    <CartDropdownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map((cartItem) => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
        )}
      </CartItemsContainer>
      <CartDropdownButton
        onClick={() => {
          user ? history.push("/checkout") : history.push("/signin");
          dispatch(toggleCartHidden());
        }}
      >
        {user ? "GO TO CHECKOUT" : "SignIN"}
      </CartDropdownButton>
    </CartDropdownContainer>
  );  
};

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

export default withRouter(connect(mapStateToProps)(CartDropdown));
