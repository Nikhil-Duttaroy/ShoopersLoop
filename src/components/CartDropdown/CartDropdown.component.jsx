import React from 'react'
import { connect } from "react-redux";
import CustomButton from './../CustomButton/CustomButton.components';
import CartItem from "./../CartItem/CartItem.component";

import { selectCartItems } from "./../../redux/cart/cart.selector";

import './CartDropdown.styles.scss'



const CartDropdown = ({ cartItems }) => (
  <div className='cart-dropdown'>
    <div className='cart-items'>
      {cartItems.map(cartItem => (
        <CartItem key={cartItem.id} item={cartItem}/>
      ))}
    </div>
    <CustomButton>GO TO CHECKOUT</CustomButton>
  </div>
);

//*Removed as we are using reselect library
// const mapStateToProps =({cart:{cartItems}}) => ({
//   cartItems
// })

const mapStateToProps =(state) => ({
   cartItems:selectCartItems(state)
})

export default connect(mapStateToProps)(CartDropdown);