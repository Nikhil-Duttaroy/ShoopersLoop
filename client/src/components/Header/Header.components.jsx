import { React } from "react";
// import { Link } from "react-router-dom";
import { connect } from "react-redux";

// import "./Header.styles.scss";
import CartIcon from "../CartIcon/CartIcon.components.jsx";
import CartDropdown from "./../CartDropdown/CartDropdown.component";

import { createStructuredSelector } from "reselect";

import { selectCartHidden } from "./../../redux/cart/cart.selector";
import { selectCurrentUser } from "./../../redux/user/user.selector";

import {
  ReactComponent as Logo,
} from "../../assets/shoppers-loop.svg";
import {
  ReactComponent as LogoSmall,
} from "../../assets/shoppers-loop-small.svg";
import  sun from "../../assets/sun.svg";
import moon from "../../assets/moon.svg";


//Styled Components import
import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink,
  OptionDiv,
} from "./Header.styles";
import { signOutStart } from "./../../redux/user/user.actions";

const Header = ({ currentUser, hidden, signOutStart, themeToggle, theme }) => (
  <HeaderContainer>
    <LogoContainer to='/'>
      {/* <Logo className='logo' /> */}
      {typeof window !== "undefined" ? (
        window.innerWidth >= 400 ? (
          <Logo />
        ) : (
          <LogoSmall />
        )
      ) : null}
    </LogoContainer>
    <OptionsContainer>
      <OptionLink to='/shop'>SHOP</OptionLink>
      <OptionLink to='/contact'>CONTACT</OptionLink>
      {currentUser ? (
        <OptionDiv onClick={signOutStart}> SIGN OUT</OptionDiv>
      ) : (
        <OptionLink to='/signin'>SIGNIN</OptionLink>
      )}
      <OptionLink onClick={() => themeToggle()}>
        {theme === "light" ? (
          <img src={moon} alt='dark mode' style={{ width: "1.5rem" }} />
        ) : (
          <img src={sun} alt='light mode' style={{ width: "1.5rem" }} />
        )}
      </OptionLink>

      <CartIcon />
    </OptionsContainer>
    {hidden ? null : <CartDropdown />}
  </HeaderContainer>
);

//*removed as using reselect library
// const mapStateToProps = ({user:{currentUser}, cart:{hidden}}) => ({
// currentUser: state.user.currentUser, //state.user.currentUser  ----> Root reducer [user] =>userReducer[currentUser]
//   currentUser,
//   hidden
// });

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden,
});

const mapDispatchToProps = (dispatch) => ({
  signOutStart: () => dispatch(signOutStart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
