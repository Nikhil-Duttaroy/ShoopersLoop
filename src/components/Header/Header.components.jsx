import { React } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";

import "./Header.styles.scss";
import { auth } from "../../firebase/firebase.utils";
import CartIcon from "../CartIcon/CartIcon.components.jsx";
import CartDropdown from "./../CartDropdown/CartDropdown.component";

import {createStructuredSelector} from 'reselect'

import { selectCartHidden } from "./../../redux/cart/cart.selector";
import { selectCurrentUser } from "./../../redux/user/user.selector";

import { ReactComponent as Logo } from "../../assets/crown.svg";


const Header = ({ currentUser ,hidden}) => (
  <div className='header'>
    <Link className='logo-container' to='/'>
      <Logo className='logo' />
    </Link>
    <div className='options'>
      <Link className='option' to='/shop'>
        SHOP
      </Link>
      <Link className='option' to='/contact'>
        CONTACT
      </Link>
      {currentUser ? (
        <div className='option' onClick={() => auth.signOut()}>
          {" "}
          Sign Out
        </div>
      ) : (
        <Link className='option' to='/signin'>
          SIGNIN
        </Link>
      )}
      <CartIcon />
      
    </div>
   { hidden ? null : <CartDropdown/>}
  </div>
);

//*removed as using reselect library
// const mapStateToProps = ({user:{currentUser}, cart:{hidden}}) => ({
//   // currentUser: state.user.currentUser, //state.user.currentUser  ----> Root reducer [user] =>userReducer[currentUser]
//   currentUser,
//   hidden
// });

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden
})

export default connect(mapStateToProps)(Header);
