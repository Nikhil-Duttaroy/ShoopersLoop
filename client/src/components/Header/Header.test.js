import { React } from 'react';
import { shallow } from 'enzyme';
import { Header } from './Header.components';
import { CartDropdown } from './../CartDropdown/CartDropdown.component';


describe('Header Component', () => {
    let wrapper;
    let mockSignOutStart;
    let mockThemeToggle;;

    beforeEach(()=>{
        mockSignOutStart=jest.fn();
        mockThemeToggle=jest.fn();
        const mockProps = {
          hidden: true,
          currentUser: {
            uid: '1',
          },
          signOutStart: mockSignOutStart,
          themeToggle:mockThemeToggle,
          theme:"dark"
        };
        wrapper = shallow(<Header {...mockProps} />);
    });

    it("should render Header component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe("if currentUser is present i.e uid is true", () => {
    it("should render sign out link", () => {
        expect(wrapper.find("OptionDiv").text()).toBe("SIGN OUT");
    });
    it("should call signOutStart method when link is clicked", () => {
        wrapper.find("OptionDiv").simulate("click");
        expect(mockSignOutStart).toHaveBeenCalled();
    });
    });

describe("if currentUser is null i.e uid is null", () => {
  it("should render sign in link", () => {
    const mockProps = {
      hidden: true,
      currentUser: null,
      signOutStart: mockSignOutStart,
      themeToggle: mockThemeToggle,
      theme: "dark",
    };
    const newWrapper = shallow(<Header {...mockProps} />);
    expect(newWrapper.find("OptionLink").at(2).text()).toBe("SIGN IN");
  });
});

describe("if hidden is true", () => {
  it("should not render CartDropdown", () => {
    expect(wrapper.exists(CartDropdown)).toBe(false);
  });
});

describe("themeToggle is called",() =>{
    it("should call themeToggle",() => {
        wrapper.find("OptionLink").at(2).simulate('click') 
        expect(mockThemeToggle).toHaveBeenCalled()
    })
})

});
