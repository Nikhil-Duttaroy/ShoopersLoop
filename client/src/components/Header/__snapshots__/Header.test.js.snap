// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should render Header component 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    <ForwardRef(SvgShoppersLoop) />
  </Styled(Link)>
  <styled.div>
    <OptionLink
      to="/shop"
    >
      SHOP
    </OptionLink>
    <OptionLink
      to="/contact"
    >
      CONTACT
    </OptionLink>
    <OptionDiv
      onClick={[MockFunction]}
    >
      SIGN OUT
    </OptionDiv>
    <OptionLink
      onClick={[Function]}
    >
      <img
        alt="light mode"
        src="sun.svg"
        style={
          Object {
            "width": "1.5rem",
          }
        }
      />
    </OptionLink>
    <Connect(CartIcon) />
  </styled.div>
</styled.div>
`;
